Source/Core/VideoCommon/BPMemory.h

```diff
@@ -1979,6 +1967,93 @@ struct BPS_TmemConfig
   u32 texinvalidate;
 };
 
+struct AllTexUnits;
+
+// The addressing of the texture units is a bit non-obvious.
+// This struct abstracts the complexity away.
+union TexUnitAddress
+{
+  enum class Register : unsigned int
+  {
+    SETMODE0 = 0,
+    SETMODE1 = 1,
+    SETIMAGE0 = 2,
+    SETIMAGE1 = 3,
+    SETIMAGE2 = 4,
+    SETIMAGE3 = 5,
+    SETTLUT = 6,
+    UNKNOWN = 7,
+  };
+
+  struct
+  {
+    u32 UnitIdLow : 2;
+    Register Reg : 3;
+    u32 UnitIdHigh : 1;
+    u32 _pad: 26;
+  };
+
+  TexUnitAddress() = default;
+  TexUnitAddress(u32 unit_id, u32 reg)
+      : UnitIdLow(unit_id & 3), Reg(Register(0)), UnitIdHigh(unit_id >> 2), _pad(0)
+  {
+  }
+
+  static TexUnitAddress FromBPAddress(u32 Address)
+  {
+    // Clear upper two bits (which should always be 0x80)
+    return Common::BitCast<TexUnitAddress>(Address & 0x3f);
+  }
+
+  u32 GetUnitID() const { return UnitIdLow | (UnitIdHigh << 2); }
+
+private:
+  friend AllTexUnits;
+
+  size_t GetOffset() const { return Common::BitCast<u32>(*this); }
+};
+static_assert(sizeof(TexUnitAddress) == sizeof(u32));
+
+// A view of the registers of a single TexUnit
+struct TexUnit
+{
+  TexMode0 texMode0;
+  u32 : 32;
+  u64 : 64;
+  TexMode1 texMode1;
+  u32 : 32;
+  u64 : 64;
+  TexImage0 texImage0;
+  u32 : 32;
+  u64 : 64;
+  TexImage1 texImage1;
+  u32 : 32;
+  u64 : 64;
+  TexImage2 texImage2;
+  u32 : 32;
+  u64 : 64;
+  TexImage3 texImage3;
+  u32 : 32;
+  u64 : 64;
+  TexTLUT texTlut;
+  u32 : 32;
+  u64 : 64;
+};
+static_assert(sizeof(TexUnit) == sizeof(u32) * 4 * 7);
+
+struct AllTexUnits
+{
+  std::array<u32, 8 * 8> AllRegisters;
+
+  const TexUnit& GetUnit(u32 UnitId) const
+  {
+    auto address = TexUnitAddress(UnitId, 0);
+    const u32* ptr = &AllRegisters[address.GetOffset()];
+    return *reinterpret_cast<const TexUnit*>(ptr);
+  }
+};
```

Comment:
In release builds with MSVC, `GetUnit` gets optimised out (which is mostly a good thing).  Unfortunately this means that MSVC's debugger can't use it.

I was able to work around this, but it ended up being a bit hacky and due to the anonymous bitfield issue I mentioned above the `pad` field is always visible:

<details><summary>Patch</summary>

```patch
diff --git a/Source/Core/VideoCommon/BPMemory.h b/Source/Core/VideoCommon/BPMemory.h
index f67679b81f..5a206ee5f6 100644
--- a/Source/Core/VideoCommon/BPMemory.h
+++ b/Source/Core/VideoCommon/BPMemory.h
@@ -1967,7 +1967,7 @@ struct BPS_TmemConfig
   u32 texinvalidate;
 };
 
-struct AllTexUnits;
+union AllTexUnits;
 
 // The addressing of the texture units is a bit non-obvious.
 // This struct abstracts the complexity away.
@@ -2011,6 +2011,10 @@ private:
   friend AllTexUnits;
 
   size_t GetOffset() const { return Common::BitCast<u32>(*this); }
+  static constexpr size_t ComputeOffset(u32 unit_id)
+  {
+    return ((unit_id & 4) << 3) | (unit_id & 3);
+  }
 };
 static_assert(sizeof(TexUnitAddress) == sizeof(u32));
 
@@ -2041,7 +2045,7 @@ struct TexUnit
 };
 static_assert(sizeof(TexUnit) == sizeof(u32) * 4 * 7);
 
-struct AllTexUnits
+union AllTexUnits
 {
   std::array<u32, 8 * 8> AllRegisters;
 
@@ -2051,6 +2055,25 @@ struct AllTexUnits
     const u32* ptr = &AllRegisters[address.GetOffset()];
     return *reinterpret_cast<const TexUnit*>(ptr);
   }
+
+private:
+  // For MSVC's debugger since GetUnit is optimised out in release builds
+  template <u32 UnitId>
+  struct OffsetTexUnit
+  {
+    static_assert(UnitId != 0, "Can't use 0 as sizeof(std::array<u32, 0>) != 0");
+    // u32 : sizeof(u32) * TexUnitAddress::ComputeOffset(UnitId);  // Legal but doesn't work on MSVC
+    std::array<u32, TexUnitAddress::ComputeOffset(UnitId)> pad;
+    TexUnit unit;
+  };
+  TexUnit tex0;
+  OffsetTexUnit<1> tex1;
+  OffsetTexUnit<2> tex2;
+  OffsetTexUnit<3> tex3;
+  OffsetTexUnit<4> tex4;
+  OffsetTexUnit<5> tex5;
+  OffsetTexUnit<6> tex6;
+  OffsetTexUnit<7> tex7;
 };
 static_assert(sizeof(AllTexUnits) == 8 * 8 * sizeof(u32));
 
```
</details>

Note that this also couldn't use `TexUnitAddress` directly since you can't use unions for type punning in `constexpr`, and `Common::BitCast<u32>` isn't `constexpr` either.  [C++20's `std::bit_cast`](https://en.cppreference.com/w/cpp/numeric/bit_cast) _is_ `constexpr`, so it could possibly be used there.

Reply:
I came up with a second solution that gets rid of the `pad` field in exchange for nesting `union { struct { union { struct { ... } } } }`:

<details><summary>Patch</summary>

```patch
diff --git a/Source/Core/VideoCommon/BPMemory.h b/Source/Core/VideoCommon/BPMemory.h
index f67679b81f..b51a874e01 100644
--- a/Source/Core/VideoCommon/BPMemory.h
+++ b/Source/Core/VideoCommon/BPMemory.h
@@ -4,6 +4,7 @@
 #pragma once
 
 #include <array>
+#include <cstddef>
 #include <string>
 #include <utility>
 
@@ -1967,7 +1968,7 @@ struct BPS_TmemConfig
   u32 texinvalidate;
 };
 
-struct AllTexUnits;
+union AllTexUnits;
 
 // The addressing of the texture units is a bit non-obvious.
 // This struct abstracts the complexity away.
@@ -2007,6 +2008,11 @@ union TexUnitAddress
 
   u32 GetUnitID() const { return UnitIdLow | (UnitIdHigh << 2); }
 
+  static constexpr size_t ComputeOffset(u32 unit_id)
+  {
+    return ((unit_id & 4) << 3) | (unit_id & 3);
+  }
+
 private:
   friend AllTexUnits;
 
@@ -2038,10 +2044,11 @@ struct TexUnit
   TexTLUT texTlut;
   u32 : 32;
   u64 : 64;
+  u32 unknown;
 };
-static_assert(sizeof(TexUnit) == sizeof(u32) * 4 * 7);
+static_assert(sizeof(TexUnit) == sizeof(u32) * ((4 * 7) + 1));
 
-struct AllTexUnits
+union AllTexUnits
 {
   std::array<u32, 8 * 8> AllRegisters;
 
@@ -2051,8 +2058,82 @@ struct AllTexUnits
     const u32* ptr = &AllRegisters[address.GetOffset()];
     return *reinterpret_cast<const TexUnit*>(ptr);
   }
+
+  struct
+  {
+    union
+    {
+      struct
+      {
+        TexUnit tex0;
+        u32 : 32;
+        u32 : 32;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        TexUnit tex1;
+        u32 : 32;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        u32 : 32;
+        TexUnit tex2;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        u32 : 32;
+        u32 : 32;
+        TexUnit tex3;
+      };
+    };
+    union
+    {
+      struct
+      {
+        TexUnit tex4;
+        u32 : 32;
+        u32 : 32;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        TexUnit tex5;
+        u32 : 32;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        u32 : 32;
+        TexUnit tex6;
+        u32 : 32;
+      };
+      struct
+      {
+        u32 : 32;
+        u32 : 32;
+        u32 : 32;
+        TexUnit tex7;
+      };
+    };
+  };
 };
 static_assert(sizeof(AllTexUnits) == 8 * 8 * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex0) == TexUnitAddress::ComputeOffset(0) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex1) == TexUnitAddress::ComputeOffset(1) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex2) == TexUnitAddress::ComputeOffset(2) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex3) == TexUnitAddress::ComputeOffset(3) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex4) == TexUnitAddress::ComputeOffset(4) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex5) == TexUnitAddress::ComputeOffset(5) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex6) == TexUnitAddress::ComputeOffset(6) * sizeof(u32));
+static_assert(offsetof(AllTexUnits, tex7) == TexUnitAddress::ComputeOffset(7) * sizeof(u32));
 
 // All of BP memory
 ```
</details>

What's nice about this is that you can also write `bpmem.tex.tex0` directly and it'll just work.

Reply:
Oh, I forgot that you might want to inspect this with a debugger. 

Unions with padding is a clever solution. I like it better than unrolled structs of unions of unions of structs.

It's a little annoying that `ComputeOffset` duplicates the address calculation algorithm, the entire point was to implement it in just one place. I guess we should leave a `TODO` to fix it once we upgrade to `std::bit_cast`

Reply:
An alternative to my first unions with padding solution is this:

<details><summary>Patch</summary>

```patch
diff --git a/Source/Core/VideoCommon/BPMemory.h b/Source/Core/VideoCommon/BPMemory.h
index f67679b81f..d7d5df89c4 100644
--- a/Source/Core/VideoCommon/BPMemory.h
+++ b/Source/Core/VideoCommon/BPMemory.h
@@ -1967,7 +1967,7 @@ struct BPS_TmemConfig
   u32 texinvalidate;
 };
 
-struct AllTexUnits;
+union AllTexUnits;
 
 // The addressing of the texture units is a bit non-obvious.
 // This struct abstracts the complexity away.
@@ -2011,6 +2011,10 @@ private:
   friend AllTexUnits;
 
   size_t GetOffset() const { return Common::BitCast<u32>(*this); }
+  static constexpr size_t ComputeOffset(u32 unit_id)
+  {
+    return ((unit_id & 4) << 3) | (unit_id & 3);
+  }
 };
 static_assert(sizeof(TexUnitAddress) == sizeof(u32));
 
@@ -2041,7 +2045,7 @@ struct TexUnit
 };
 static_assert(sizeof(TexUnit) == sizeof(u32) * 4 * 7);
 
-struct AllTexUnits
+union AllTexUnits
 {
   std::array<u32, 8 * 8> AllRegisters;
 
@@ -2051,6 +2055,51 @@ struct AllTexUnits
     const u32* ptr = &AllRegisters[address.GetOffset()];
     return *reinterpret_cast<const TexUnit*>(ptr);
   }
+
+private:
+  // For MSVC's debugger since GetUnit is optimised out in release builds
+  template <u32 UnitId>
+  struct TexUnitPadding
+  {
+    static_assert(UnitId != 0, "Can't use 0 as sizeof(std::array<u32, 0>) != 0");
+    std::array<u32, TexUnitAddress::ComputeOffset(UnitId)> pad;
+  };
+  TexUnit tex0;
+  struct
+  {
+    TexUnitPadding<1> pad1;
+    TexUnit tex1;
+  };
+  struct
+  {
+    TexUnitPadding<2> pad2;
+    TexUnit tex2;
+  };
+  struct
+  {
+    TexUnitPadding<3> pad3;
+    TexUnit tex3;
+  };
+  struct
+  {
+    TexUnitPadding<4> pad4;
+    TexUnit tex4;
+  };
+  struct
+  {
+    TexUnitPadding<5> pad5;
+    TexUnit tex5;
+  };
+  struct
+  {
+    TexUnitPadding<6> pad6;
+    TexUnit tex6;
+  };
+  struct
+  {
+    TexUnitPadding<7> pad7;
+    TexUnit tex7;
+  };
 };
 static_assert(sizeof(AllTexUnits) == 8 * 8 * sizeof(u32));
```
</details>

The benefit is that `tex1` and the like are directly part of `bpmem.tex` instead of needing to access it via `bpmem.tex1.unit`; this is much more convenient when debugging because it means that the watch window doesn't cut it off as much.